feat(#1699): new 4.x upgrade troubleshooting guide (#1709)
medic/cht-docs: content/en/hosting/4.x/upgrade-troubleshooting.md

---
title: "Troubleshooting 4.x upgrades"
linkTitle: "Troubleshooting upgrades"
weight: 50
aliases:
  - 
description: >
  What to do when CHT 4.x upgrades don't work as planned
relatedContent: >
  hosting/4.x/data-migration
---

> 4.0.0 was released in November of 2022 so 4.x is now well into a mature and Medic has learned a number of important lessons on how to unstick 4.x upgrades that get stuck.  Below are some specific tips as well as general practices on upgrading 4.x.

Reviewer: ```suggestion
4.0.0 was released in November of 2022 so 4.x is mature and users has learned a number of important lessons on how to unstick 4.x upgrades that get stuck.  Below are some specific tips as well as general practices on upgrading 4.x.
```

{{% pageinfo %}}
All tips apply to both [Docker]({{< relref "hosting/4.x/production/docker" >}}) and [Kubernetes]({{< relref "hosting/4.x/production/kubernetes" >}}) based deployments unless otherwise specified.

All upgrades are expected to succeed without issue.  Do not attempt any fixes unless you actively have a problem upgrading.
{{% /pageinfo %}}

## Considerations

> When troubleshooting, consider make sure there are:

Reviewer: ```suggestion
When troubleshooting, consider making sure there are:
```

* Backups exist and restores have been tested 
* Extra disk space is availabe (up to 5x!)
* The upgrade has been tested on a development instance with production data

## A go-to fix: restart

> A safe fix for any upgrade getting stuck is to restart all services.  Any views that were being re-indexed will be picked up where they left off without loosing any work.  This should be your first step when trouble shooting a stuck upgrade. 

Reviewer: should we describe the symptoms of what you mean that an upgrade "gets stuck"?

Author: good thinking! since we reference it many times, I moved the definition toward the top of the page in [d6b1228](https://github.com/medic/cht-docs/pull/1709/commits/d6b122820388988630a5a12f030a280ead35e2bd)

> A safe fix for any upgrade getting stuck is to restart all services.  Any views that were being re-indexed will be picked up where they left off without loosing any work.  This should be your first step when trouble shooting a stuck upgrade. 

Reviewer: ```suggestion
A safe fix for any upgrade getting stuck is to restart all services.  Any views that were being re-indexed will be picked up where they left off without losing any work.  This should be your first step when trouble shooting a stuck upgrade. 
```

```

CouchDB logs scenario 2:

```shell
[info] 2024-11-04T20:18:46.692239Z couchdb@127.0.0.1 <0.6832.4663> -------- Starting compaction for db "shards/7ffffffe-95555552/medic-user-mikehaya-meta.1690191139" at 10
[info] 2024-11-04T20:19:47.821999Z couchdb@127.0.0.1 <0.7017.4653> -------- Starting compaction for db "shards/7ffffffe-95555552/medic-user-marnyakoa-meta.1690202463" at 21
[info] 2024-11-04T20:21:24.529822Z couchdb@127.0.0.1 <0.24125.4661> -------- Starting compaction for db "shards/7ffffffe-95555552/medic-user-lilian_lubanga-meta.1690115504" at 15
```

**Fix:** Give CouchDB more disk and Restart all services

> _* See eCHIS Kenya [Issue #2578](https://github.com/moh-kenya/config-echis-2.0/issues/2578#issuecomment-2455702112) - a private repo and not available to the public_

Reviewer: is this useful? linking to private issues? who is this for if we are linking to these?

Author: Fair!  

That said, I think it's worth keeping. In the very short term this will help the folks that directly worked on it. Months from now it could help other Medic and eCHIS staff that still have access to these private repos.  Agreed quarters/years from now they'll be useless, but so will much of this guide by then (I hope!).

i vote to keep as possibly helpful for a select few, so worth it.  gently resolving, but feel free to re-open if you feel strongly!

---
title: "Troubleshooting 4.x upgrades"
linkTitle: "Troubleshooting upgrades"
weight: 50
aliases:
  - 
description: >
  What to do when CHT 4.x upgrades don't work as planned
relatedContent: >
  hosting/4.x/data-migration
---

4.0.0 was released in November of 2022 so 4.x is mature and users have learned a number of important lessons on how to fix failed 4.x upgrades.  Below are some specific tips as well as general practices on upgrading 4.x.

> There's a concept of upgrades "getting stuck" which mainly means that after many many hours an upgrade is not making any progress.  Most likely, this will manifest as the progress bars in the  upgrade admin web UI not increasing and "sticking" at a certain percentage.

Reviewer 2: This can also mean the progress bars disappearing altogether.

Author: gotcha! fixed in 6fcbd395

**[Issue #9286](https://github.com/medic/cht-core/issues/9286)**:  Starting an upgrade that involves view indexing can cause CouchDB to crash on large databases (>30m docs).  The upgrade will fail and you will see the logs below when you have this issue.

HAProxy:

```shell
<150>Jul 26 20:57:39 haproxy[12]: 172.22.0.4,<NOSRV>,503,0,0,0,GET,/,-,medic,'-',217,-1,-,'-'
<150>Jul 26 20:57:40 haproxy[12]: 172.22.0.4,<NOSRV>,503,0,0,0,GET,/,-,medic,'-',217,-1,-,'-'
<150>Jul 26 20:57:41 haproxy[12]: 172.22.0.4,<NOSRV>,503,0,0,0,GET,/,-,medic,'-',217,-1,-,'-'
```

CouchDB

```shell
[notice] 2024-07-26T20:52:45.229027Z couchdb@127.0.0.1 <0.10998.4> dca1387a05 haproxy:5984 172.22.0.4 medic GET /_active_tasks 200 ok 1
[notice] 2024-07-26T20:52:45.234397Z couchdb@127.0.0.1 <0.10986.4> 2715cd1e47 haproxy:5984 172.22.0.4 medic GET /medic-logs/_all_docs?descending=true&include_docs=true&startkey=%22upgrade_log%3A1722027165223%3A%22&limit=1 200 ok 6
[notice] 2024-07-26T20:52:45.468469Z couchdb@127.0.0.1 <0.11029.4> 3bf85c6071 haproxy:5984 172.22.0.4 medic GET /_active_tasks 200 ok 1
```

**Fix:**
1. Check that all the indexes are warmed by loading them one by one in fauxton.

> 2. Restart all services, **retry** upgrade from Admin GUI - do not cancel and upgrade.

Reviewer 2: Maybe we can include an image of the retry button to be super clear

Author: yes! i added on in f0ae84fc . I wasn't sure exactly how to recreate that locally, so it's a real screenshot, but slightly doctored version strings to meet our use case.  Lemme know if it's not accurate!

see f0ae84fc

## CHT 4.0.0 - 4.2.2: view indexing can become stuck after indexing is finished

**[Issue #9617](https://github.com/medic/cht-core/issues/9617):** Starting an upgrade that involves view indexing can become stuck after indexing is finished

Upgrade process stalls while trying to index staged views:

![CHT Core admin UI showing upgrade progress bar stalled at 4% ](stalled-upgrade.png)

**Fix:**

> Unfortunately, the workaround is manual and very technical and involves:

Reviewer 2: > the workaround is manual and very technical and involves:

I believe this remained as a statement from the time when we instructed to manually edit upgrade logs. Is this "very technical"?

Author: fixed in fc5f2faf - thanks!

 

## CHT 4.0.1 - 4.9.0: CouchDB restart causes all services to go down

**Note** - This is a Docker only issue.

**[issue](https://github.com/medic/cht-core/issues/9284)**:   A couchdb restart in single node docker takes down the whole instance.

Haproxy continuously reports NOSRV errors like:

```shell
<150>Jul 25 18:11:03 haproxy[12]: 172.18.0.9,<NOSRV>,503,0,1001,0,GET,/,-,admin,'-',241,-1,-,'-'
```

API logs:

```shell
StatusCodeError: 503 - {"error":"503 Service Unavailable","reason":"No server is available to handle this request","server":"haproxy"}
```

nginx reports:

```shell
2024/07/25 18:40:28 [error] 43#43: *5757 connect() failed (111: Connection refused) while connecting to upstream, client: 172.18.0.1, 
```

**Fix:** Restart all services

## CHT 4.2.4 - 4.3.x: CouchDB Crashes

**[issue](https://github.com/medic/cht-core/issues/9617):** Starting an upgrade that involves view indexing can become stuck after indexing is finished

upgrade process stalls after view indexes are built

**Fix:**

Unfortunately, the workaround is manual and very technical and involves:

> * identifying the upgrade log document in the medic-logs db.

Reviewer 2: I have shared a new less technical workaround for this on the issue: https://github.com/medic/cht-core/issues/9617#issuecomment-2465299251

1. When API goes stuck after view indexing, simply restart API.
2. The admin upgrade page will say that the upgrade was interrupted, click retry upgrade.
3. Depending on the state of the database, you might see view indexing again. Depending on how many docs need to be indexed, indexing might get stuck again. Go back to 1 if that happens.
4. Eventually, when indexing jobs are short enough not to trigger a request hang, you will get the button to complete the upgrade.

Author: thanks! should be fixed in a745faab

   The following pods are not ready...."}
```

Running `kubectl get po` shows 3 pods with status of `ContainerStatusUnknown`:

![CLI screenshot showing 3 pods with STATUS of "ContainerStatusUnknown" ](container-status-unknown.png)

**Fix:** delete pods so they get recreated and start cleanly 

(tk - is this syntax legal/correct?)

`kubectl delete po 'cht.service in (api, sentinel, haproxy, couchdb)'`

## 4.2 -> 4.11 upgrades when stuck in "indexing"

Issue: ?? just slack thread on ticket https://github.com/medic/cht-docs/issues/1699

**Fix:**
1. I'm checking that all the indexes are warmed by loading them one by one in fauxton.
1. If all views load, then I'm going to medic-logs database and finding the latest upgrade_log
1. I'm setting the state from `indexing` to `indexed` and creating a fake section in `state_history`

---- OR IS IT THIS SOLUTION?! ----

1. When API goes stuck after view indexing, restart
1. The admin upgrade page will say that the upgrade was interrupted, click retry upgrade.
1. Depending on the state of the database, you might see view indexing again. Depending on how many docs need to be indexed, indexing might get stuck again. Go back to 1 if that happens.

> 1. Eventually, when indexing jobs are short enough not to trigger a request hang, you will get the button to complete the upgrade.

Reviewer 2: Yes, this is the solution.
